File: typescript/prototype/manager/Manager.ts
import { Product } from "./Product.ts";

export class Manager {
  private showcase: { [key: string]: Product };

  constructor() {
    this.showcase = {};
  }

  public register = (name: string, proto: Product) => {
    this.showcase[name] = proto;
  };

  public create = (name: string): Product => {
    const prd: Product = Object.assign(
      Object.create(
        Object.getPrototypeOf(this.showcase[name]),
      ),
      this.showcase[name],
    );

    return prd;
  };
}

File: go/prototype/manager/manager.go
package manager

type Product interface {
	Use(string) string
}

type Manager struct {
	showcase map[string]Product
}

func New() *Manager {
	return &Manager{map[string]Product{}}
}

func (m *Manager) Register(key string, p Product) {
	m.showcase[key] = p
}

func (m *Manager) Create(key string) Product {
	copy := m.showcase[key]
	return copy
}
